diff --git a/bin/swift-dispersion-populate b/bin/swift-dispersion-populate
index 6e208cc..4f60d80 100755
--- a/bin/swift-dispersion-populate
+++ b/bin/swift-dispersion-populate
@@ -91,19 +91,24 @@ if __name__ == '__main__':
     dispersion_coverage = int(conf.get('dispersion_coverage', 1))
     retries = int(conf.get('retries', 5))
     concurrency = int(conf.get('concurrency', 25))
+    endpoint_type = str(conf.get('endpoint_type', 'publicURL'))
 
     coropool = GreenPool(size=concurrency)
     retries_done = 0
 
+    os_options = {'endpoint_type': endpoint_type}
+
     url, token = get_auth(conf['auth_url'], conf['auth_user'],
                           conf['auth_key'],
-                          auth_version=conf.get('auth_version', '1.0'))
+                          auth_version=conf.get('auth_version', '1.0'),
+                          os_options=os_options)
     account = url.rsplit('/', 1)[1]
     connpool = Pool(max_size=concurrency)
     connpool.create = lambda: Connection(conf['auth_url'],
                                 conf['auth_user'], conf['auth_key'],
                                 retries=retries,
-                                preauthurl=url, preauthtoken=token)
+                                preauthurl=url, preauthtoken=token,
+                                os_options=os_options)
 
     container_ring = Ring(swift_dir, ring_name='container')
     parts_left = dict((x, x) for x in xrange(container_ring.partition_count))
diff --git a/bin/swift-dispersion-report b/bin/swift-dispersion-report
index a055016..8dd76f1 100755
--- a/bin/swift-dispersion-report
+++ b/bin/swift-dispersion-report
@@ -320,6 +320,7 @@ Usage: %prog [options] [conf_file]
     dispersion_coverage = int(conf.get('dispersion_coverage', 1))
     retries = int(conf.get('retries', 5))
     concurrency = int(conf.get('concurrency', 25))
+    endpoint_type = str(conf.get('endpoint_type', 'publicURL'))
     if options.dump_json or config_true_value(conf.get('dump_json', 'no')):
         json_output = True
     if options.debug:
@@ -327,14 +328,17 @@ Usage: %prog [options] [conf_file]
 
     coropool = GreenPool(size=concurrency)
 
+    os_options = {'endpoint_type': endpoint_type}
+
     url, token = get_auth(conf['auth_url'], conf['auth_user'],
                           conf['auth_key'],
-                          auth_version=conf.get('auth_version', '1.0'))
+                          auth_version=conf.get('auth_version', '1.0'),
+                          os_options=os_options)
     account = url.rsplit('/', 1)[1]
     connpool = Pool(max_size=concurrency)
     connpool.create = lambda: Connection(
         conf['auth_url'], conf['auth_user'], conf['auth_key'], retries=retries,
-        preauthurl=url, preauthtoken=token)
+        preauthurl=url, preauthtoken=token, os_options=os_options)
 
     container_ring = Ring(swift_dir, ring_name='container')
     object_ring = Ring(swift_dir, ring_name='object')
diff --git a/doc/manpages/swift-dispersion-populate.1 b/doc/manpages/swift-dispersion-populate.1
index 27188ce..e9cb0b2 100644
--- a/doc/manpages/swift-dispersion-populate.1
+++ b/doc/manpages/swift-dispersion-populate.1
@@ -69,6 +69,7 @@ Example \fI/etc/swift/dispersion.conf\fR:
 .IP "# dispersion_coverage = 1"
 .IP "# retries = 5"
 .IP "# concurrency = 25"
+.IP "# endpoint_type = publicURL"
 .RE
 .PD 
 .SH EXAMPLE
diff --git a/doc/manpages/swift-dispersion-report.1 b/doc/manpages/swift-dispersion-report.1
index 414e280..72cb7e0 100644
--- a/doc/manpages/swift-dispersion-report.1
+++ b/doc/manpages/swift-dispersion-report.1
@@ -86,6 +86,7 @@ Example \fI/etc/swift/dispersion.conf\fR:
 .IP "# retries = 5"
 .IP "# concurrency = 25"
 .IP "# dump_json = no"
+.IP "# endpoint_type = publicURL"
 .RE
 .PD 
 .SH EXAMPLE
diff --git a/etc/dispersion.conf-sample b/etc/dispersion.conf-sample
index 79ceac6..8a0bc12 100644
--- a/etc/dispersion.conf-sample
+++ b/etc/dispersion.conf-sample
@@ -6,6 +6,7 @@ auth_key = testing
 # auth_user = test:tester
 # auth_key = testing
 # auth_version = 2.0
+# endpoint_type = publicURL
 #
 # swift_dir = /etc/swift
 # dispersion_coverage = 1
