Description: Fix keystoneauth with OPTIONS headers (eg: CORS)
Author: Chmouel Boudjnah <chmouel@chmouel.com>
Origin: https://review.openstack.org/#/c/20017/
Bug-Ubuntu: https://bugs.launchpad.net/swift/+bug/1101143

diff --git a/swift/common/middleware/keystoneauth.py b/swift/common/middleware/keystoneauth.py
index b7cdd34..cef55bf 100644
--- a/swift/common/middleware/keystoneauth.py
+++ b/swift/common/middleware/keystoneauth.py
@@ -107,6 +107,11 @@ class KeystoneAuth(object):
             self.logger.debug(msg)
             return self.app(environ, start_response)
 
+        #allow OPTIONS requests to proceed as normal
+        if ('REQUEST_METHOD' in environ and
+                environ['REQUEST_METHOD'] == 'OPTIONS'):
+            return self.app(environ, start_response)
+
         if identity:
             self.logger.debug('Using identity: %r' % (identity))
             environ['keystone.identity'] = identity
diff --git a/test/unit/common/middleware/test_keystoneauth.py b/test/unit/common/middleware/test_keystoneauth.py
index a40898b..d4fa192 100644
--- a/test/unit/common/middleware/test_keystoneauth.py
+++ b/test/unit/common/middleware/test_keystoneauth.py
@@ -49,7 +49,7 @@ class SwiftAuth(unittest.TestCase):
         return Request.blank(path, headers=headers, **kwargs)
 
     def _get_identity_headers(self, status='Confirmed', tenant_id='1',
-                          tenant_name='acct', user='usr', role=''):
+                              tenant_name='acct', user='usr', role=''):
         return dict(X_IDENTITY_STATUS=status,
                     X_TENANT_ID=tenant_id,
                     X_TENANT_NAME=tenant_name,
@@ -113,6 +113,12 @@ class SwiftAuth(unittest.TestCase):
         resp = req.get_response(self.test_auth)
         self.assertEquals(resp.status_int, 404)
 
+    def test_options_allowed(self):
+        req = self._make_request('/v1/AUTH_account',
+                                 environ={'REQUEST_METHOD': 'OPTIONS'})
+        resp = req.get_response(self.test_auth)
+        self.assertEquals(resp.status_int, 404)
+
 
 class TestAuthorize(unittest.TestCase):
     def setUp(self):
-- 
1.7.9.5

